pkg/controller/schedulingmanager/controller.go

```diff
@@ -152,76 +189,59 @@ func (c *SchedulerController) reconcile(qualifiedName util.QualifiedName) util.R
 	}
 
 	// Scheduling preference controller is started on demand
-	schedulingKind := schedulingType.Kind
-	scheduler, ok := c.scheduler[schedulingKind]
+	scheduler, ok := c.schedulers[schedulingKind]
 	if !ok {
-		var err error
-
-		scheduler, err = schedulingpreference.StartSchedulingPreferenceController(c.config, *schedulingType)
+		glog.Infof("Starting schedulingpreference controller for %s", schedulingKind)
+		stopChan := make(chan struct{})
+		schedulerInterface, err := schedulingpreference.StartSchedulingPreferenceController(c.config, *schedulingType, stopChan)
 		if err != nil {
-			runtime.HandleError(errors.Wrapf(err, "Error starting schedulingpreference controller for %q", schedulingKind))
+			runtime.HandleError(errors.Wrapf(err, "Error starting schedulingpreference controller for %s", schedulingKind))
 			return util.StatusError
 		}
-		c.scheduler[schedulingKind] = scheduler
+		scheduler = newSchedulerWrapper(schedulerInterface, stopChan)
+		c.schedulers[schedulingKind] = scheduler
 	}
 
-	federatedKind := typeConfig.GetFederatedType().Kind
-	glog.Infof("Starting plugin kind %s for scheduling type %s", federatedKind, schedulingKind)
+	if scheduler.HasPlugin(typeConfigName) {
+		// Scheduler and plugin already running for this target typeConfig
+		return util.StatusAllOK
+	}
 
+	federatedKind := typeConfig.GetFederatedType().Kind
+	glog.Infof("Starting plugin %s for %s", federatedKind, schedulingKind)
 	err = scheduler.StartPlugin(typeConfig)
 	if err != nil {
-		runtime.HandleError(errors.Wrapf(err, "Error starting plugin for %q", federatedKind))
+		runtime.HandleError(errors.Wrapf(err, "Error starting plugin %s for %s", federatedKind, schedulingKind))
 		return util.StatusError
 	}
-	c.runningPlugins.Insert(qualifiedName.Name)
-	c.federatedKindMap[qualifiedName.Name] = federatedKind
+	scheduler.pluginMap.Store(typeConfigName, federatedKind)
 
 	return util.StatusAllOK
 }
 
-func (c *SchedulerController) stopScheduler(schedulingKind string, qualifiedName util.QualifiedName) {
+func (c *SchedulingManager) stopScheduler(schedulingKind, typeConfigName string) {
 	c.Lock()
 	defer c.Unlock()
-	scheduler, ok := c.scheduler[schedulingKind]
+	scheduler, ok := c.schedulers[schedulingKind]
 	if !ok {
 		return
 	}
 
-	if !c.runningPlugins.Has(qualifiedName.Name) {
+	// This typeConfig might not have an associated plugin but another type could.
+	if !scheduler.HasPlugin(typeConfigName) && scheduler.pluginMap.Size() > 0 {
```

Comment:
It seems no need to add `scheduler.pluginMap.Size() > 0`

It seems try to cover the case that `!scheduler.HasPlugin(typeConfigName)` and `scheduler.pluginMap.Size() == 0`. If that's true, the later logic for handling plugin seems make no sense, right?

Reply:
Nope, its covering the case `!scheduler.HasPlugin(typeConfigName)` and `scheduler.pluginMap.Size() > 0` exactly as in code. In other words this stop is not for the plugin associated with this typeConfig, but there is/are some other plugin(s) still associated with this scheduler.
Looking at your query I figured I made the condition check unnecessarily complex. A better way of expressing the same thing is the updated code. PTAL.

Reply:
The result is what I expected. Thanks :-)